### PR TITLE
Random max in mutable special

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
Infrastructure "Support randomized max values in mutable overmap special placement rules"

#### Purpose of change
Prior to this change mutable overmap specials only supported a fixed maximum number of instance of a particular rule being used in any phase.  This meant that all instances will be of roughly the same size.  For example, all anthills had essentially the same number of tunnel tiles.

It would be nice to permit some variation.

#### Describe the solution
Allow the max to be drawn from a Poisson distribution.

Use this feature in anthills to make them a little more variable.

#### Describe alternatives you've considered
Adding more distributions.  That will be easy to do later now that the infrastructure is in place.

#### Testing
Anthill spawning unit test is still passing.  Spawned some anthills in-game with the results pictured below.

#### Additional context
Here are some anthills I spawned.  They should be of different sizes now, although I realise that's not trivial to see.
varisized-anthills

### CODE REVIEW

**Comment by actual-nh** [comment]
Should "max" be respected for the final phase?

File: src/overmap.cpp [hunk 1]
```diff
@@ -1092,13 +1093,23 @@ struct mutable_overmap_placement_rule {
...
+struct mutable_overmap_placement_rule {
+    std::string overmap;
+    int_distribution max = int_distribution( INT_MAX );
+    int weight = INT_MAX;
+
+    mutable_overmap_placement_rule_remainder realise() const {
+        return { overmap, max.sample(), weight };
+    }
 
     void deserialize( JsonIn &jin ) {
```
Comment (actual-nh) [inline src/overmap.cpp:1107]: Note #51449, BTW.

Comment (jbytheway) [reply src/overmap.cpp:1107]: Cool; good to see that merged already.  I will update accordingly.

**Comment by jbytheway** [comment]
> Should "max" be (fully) respected for the final phase?

I can see the advantage to outlawing it.  It could be used correctly in the final phase, but more likely you wouldn't want to.  And if you really wanted to I guess you could work around that by adding a fake final phase.  I'm leaning towards outlawing it as you suggested.  (But not in this PR)

**Comment by actual-nh** [comment]
> > Should "max" be (fully) respected for the final phase?
> 
> I can see the advantage to outlawing it. It could be used correctly in the final phase, but more likely you wouldn't want to. And if you really wanted to I guess you could work around that by adding a fake final phase. I'm leaning towards outlawing it as you suggested. (But not in this PR)

(No problem re not in this PR.) Perhaps allow "max" but never decrease it to 0 (minimum 1) for the final phase?

**Comment by jbytheway** [comment]
Rebased to hopefully fix the clang-tidy errors and switch to new serialization framework.

File: src/distribution.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#ifndef CATA_SRC_DISTRIBUTION_H
+#define CATA_SRC_DISTRIBUTION_H
+
+#include "memory_fast.h"
+
+struct int_distribution_impl;
+class JsonValue;
+
+// This represents a probability distribution over the integers, which is
+// abstract and can be reqad from a JSON definition
```
Comment (anothersimulacrum) [inline src/distribution.h:10]:
```suggestion
// abstract and can be read from a JSON definition
```

**Comment by jbytheway** [comment]
> (No problem re not in this PR.) Perhaps allow "max" but never decrease it to 0 (minimum 1) for the final phase?

I don't like that idea.  Inconsistent behaviour will just cause confusion.

**Pull request merged**